monasca_rsyslog/client.py

```diff
@@ -46,13 +59,67 @@ def __init__(self):
             user_agent='rsyslog-monasca'
         )
         self._url = cfg.CONF.api.url
+        self._verbose = cfg.CONF.api.verbose
+        self._min_poll_delay = cfg.CONF.api.min_poll_delay
+        self._max_batch_size = cfg.CONF.api.max_batch_size
 
-    def post_logs(self, data):
-        """Post logs to Monasca which are suitably pre-encoded as JSON."""
-        
-        self._sess.post(
-            '/logs',
-            endpoint_override=self._url,
-            headers={ 'Content-Type': 'application/json' },
-            data=data
-        )
+    def combine_logs(self, data, log_buffer):
+        """ Combine with the existing log buffer if not empty. """
+
+        log_count = 0
+        if data != None:
+            if self._verbose:
+                print(data)
+                sys.stdout.flush()
+            for key, value in jsonutils.loads(data).items():
+                log_buffer.setdefault(key, []).extend(value)
+                log_count += len(value)
+        else:
+            if self._verbose:
+                print("No input from stdin for {} seconds.".format(self._min_poll_delay))
+                sys.stdout.flush()
+        return log_count
+
+    def retry_post_logs(self, log_buffer):
+        """ Try to post logs forever until success. """
+
+        while True:
+            try:
+                # Try to post the things in the buffer
+                self._sess.post(
+                    '/logs',
+                    endpoint_override=self._url,
+                    headers={ 'Content-Type': 'application/json' },
+                    data=jsonutils.dumps(log_buffer)
+                )
+                break
+            except Exception as e:
+                # If there is a failure, most likely ConnectionFailure,
+                # write to stdout and stderr then try again.
+                print("Connection failure, trying again.")
+                sys.stdout.flush()
+                sys.stderr.write(e)
+                sys.stderr.flush()
+                time.sleep(1)
```

Comment:
Good idea putting this in - if the API goes down this will prevent vast amounts of `connection failed` log spam.